crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3431,6 +3452,139 @@ impl<'db> TypeInferenceBuilder<'db> {
+                    | ruff_python_ast::Expr::BooleanLiteral(_)
+                    // For enum values
+                    | ruff_python_ast::Expr::Attribute(_)
+                    // For Another Literal inside this Literal
+                    | ruff_python_ast::Expr::Subscript(_)
+                    | ruff_python_ast::Expr::NoneLiteral(_) => {}
+                    // for negative numbers
+                    ruff_python_ast::Expr::UnaryOp(ref u) if (u.op == UnaryOp::USub || u.op == UnaryOp::UAdd) && u.operand.is_number_literal_expr() => {}
+                    ruff_python_ast::Expr::NumberLiteral(ref number) if number.value.is_int() => {}
+                    ruff_python_ast::Expr::Tuple(ref t) if !t.parenthesized => {}
+                    _ => {
+                        self.add_diagnostic(
+                            parameters.into(),
+                            "invalid-literal-parameter",
+                            format_args!(
+                                "Type arguments for `Literal` must be None, a literal value (int, bool, str, or bytes), or an enum value",
+                            ),
+                        );
+                        return Type::Unknown;
+                    }
+                };
+
+                let slice_ty = self.infer_literal_parameter_type(parameters);
+
+                match slice_ty {
+                    Type::Never
+                    | Type::Unknown
+                    | Type::Unbound
+                    | Type::Todo
+                    | Type::FunctionLiteral(_)
+                    | Type::ModuleLiteral(_)
+                    | Type::ClassLiteral(_)
+                    | Type::Union(_)
+                    | Type::Intersection(_)
+                    | Type::Any => {
+                        self.add_diagnostic(
+                                    parameters.into(),
+                                    "invalid-literal-parameter",
+                                    format_args!(
+                                        "Type arguments for `Literal` must be None, a literal value (int, bool, str, or bytes), or an enum value",
+                                    ),
+                                );
+                        Type::Unknown
+                    }
+                    Type::Tuple(tuple) => {
+                        let elts = tuple.elements(self.db);
+                        Type::Union(UnionType::new(self.db, elts))
+                    }
+                    ty => ty,
+                }
+            }
+        }
+    }
+
+    fn infer_literal_parameter_type(&mut self, parameters: &ast::Expr) -> Type<'db> {
+        // If the slice itself is another subscript and it is another Literal then flatten
+        // the values
+        match parameters {
+            ruff_python_ast::Expr::Subscript(inner_literal_subscript) => {
+                let inner_subscript_value = self.infer_expression(&inner_literal_subscript.value);
```

Comment:
Thanks. Do you think it's useful to add a function that can detect if a type is any of the Literal types? I put it here since it's only here at the momnent.